File: packages/core/src/edit.ts
/**
 * Edit-mode machinery (EditableLabel, EditableList).
 * Contracts: docs/contracts/components/editable-label.md, editable-list.md,
 * "Behavior Machine" sections.
 *
 * Edit-label machine: view/editing with portable set-T commit trim,
 * Unicode-scalar maxLength, and cancel-restores semantics. List grab-and-move:
 * pure keyboard-intent resolution for accessible reordering (grab, arrow-move,
 * boundary, escape-cancel). Focus calls, announcements, and drag DOM plumbing
 * stay adapter-side.
 */

import type { TransitionResult } from "./machine";

// ── Portable set-T trim (Unicode White_Space + U+FEFF) ──

const EDITABLE_LABEL_TRIM_SCALARS: ReadonlySet<number> = new Set([
  0x0009, 0x000a, 0x000b, 0x000c, 0x000d, 0x0020, 0x0085, 0x00a0, 0x1680, 0x2000, 0x2001, 0x2002,
  0x2003, 0x2004, 0x2005, 0x2006, 0x2007, 0x2008, 0x2009, 0x200a, 0x2028, 0x2029, 0x202f, 0x205f,
  0x3000, 0xfeff,
]);

function isTrimScalar(char: string): boolean {
  const codePoint = char.codePointAt(0);
  return codePoint !== undefined && EDITABLE_LABEL_TRIM_SCALARS.has(codePoint);
}

/** Drop the longest prefix and suffix in portable set T. Interior scalars stay. */
export function trimEditableLabel(value: string): string {
  const chars = [...value];
  let start = 0;
  let end = chars.length;

  while (start < end && isTrimScalar(chars[start] as string)) {
    start += 1;
  }

  while (end > start && isTrimScalar(chars[end - 1] as string)) {
    end -= 1;
  }

  return chars.slice(start, end).join("");
}

/** Keep at most `maxLength` Unicode scalar values. */
export function clampEditableLabelDraft(draft: string, maxLength: number | null): string {
  if (maxLength === null || !Number.isFinite(maxLength) || maxLength < 0) {
    return draft;
  }

  const limit = Math.floor(maxLength);
  const chars = [...draft];
  return chars.length <= limit ? draft : chars.slice(0, limit).join("");
}

// ── Edit-label machine ──

export type EditLabelState = "view" | "editing";

export interface EditLabelContext {
  value: string;
  draft: string;
  disabled: boolean;
  maxLength: number | null;
}

export type EditLabelEvent =
  | { type: "START_EDIT" }
  | { type: "SET_DRAFT"; draft: string }
  | { type: "COMMIT" }
  | { type: "COMMIT_BLUR" }
  | { type: "CANCEL" }
  | { type: "REPLACE_VALUE"; value: string }
  | { type: "SET_DISABLED"; disabled: boolean }
  | { type: "TEARDOWN" };

export type EditLabelEffect =
  | { type: "emitEditStart" }
  | { type: "focusInput" }
  | { type: "emitCommit"; value: string; previousValue: string; restoreFocus: boolean }
  | { type: "emitCancel"; restoreFocus: boolean };

export type EditLabelResult = TransitionResult<EditLabelState, EditLabelContext, EditLabelEffect>;

export function editLabelContext(init: Partial<EditLabelContext> = {}): EditLabelContext {
  return {
    value: "",
    draft: "",
    disabled: false,
    maxLength: null,
    ...init,
  };
}

function commitFromEditing(
  context: EditLabelContext,
  restoreFocus: boolean,
): EditLabelResult {
  const value = trimEditableLabel(context.draft);

  return {
    state: "view",
    context: { ...context, draft: context.value },
    effects: [{ type: "emitCommit", value, previousValue: context.value, restoreFocus }],
  };
}

function cancelFromEditing(context: EditLabelContext, restoreFocus: boolean): EditLabelResult {
  return {
    state: "view",
    context: { ...context, draft: context.value },
    effects: [{ type: "emitCancel", restoreFocus }],
  };
}

export function editLabelTransition(
  state: EditLabelState,
  context: EditLabelContext,
  event: EditLabelEvent,
): EditLabelResult {
  const stay: EditLabelResult = { state, context, effects: [] };

  switch (event.type) {
    case "START_EDIT": {
      if (state !== "view" || context.disabled) {
        return stay;
      }

      return {
        state: "editing",
        context: { ...context, draft: context.value },
        effects: [{ type: "emitEditStart" }, { type: "focusInput" }],
      };
    }
    case "SET_DRAFT": {
      if (state !== "editing") {
        return stay;
      }

      const draft = clampEditableLabelDraft(event.draft, context.maxLength);
      return { state, context: { ...context, draft }, effects: [] };
    }
    case "COMMIT": {
      return state === "editing" ? commitFromEditing(context, true) : stay;
    }
    case "COMMIT_BLUR": {
      return state === "editing" ? commitFromEditing(context, false) : stay;
    }
    case "CANCEL": {
      return state === "editing" ? cancelFromEditing(context, true) : stay;
    }
    case "REPLACE_VALUE": {
      if (state !== "editing") {
        return {
          state,
          context: { ...context, value: event.value, draft: event.value },
          effects: [],
        };
      }

      if (event.value === context.value) {
        return stay;
      }

      return {
        state: "view",
        context: { ...context, value: event.value, draft: event.value },
        effects: [],
      };
    }
    case "SET_DISABLED": {
      const next = { ...context, disabled: event.disabled };

      if (state === "editing" && event.disabled) {
        return {
          state: "view",
          context: { ...next, draft: context.value },
          effects: [{ type: "emitCancel", restoreFocus: false }],
        };
      }

      return { state, context: next, effects: [] };
    }
    case "TEARDOWN": {
      if (state !== "editing") {
        return stay;
      }

      return {
        state: "view",
        context: { ...context, draft: context.value },
        effects: [],
      };
    }
  }
}

// ── List grab-and-move keyboard intents ──

export type ListReorderIntent =
  | { type: "grab" }
  | { type: "drop" }
  | { type: "cancelGrab" }
  | { type: "move"; from: number; to: number }
  | { type: "boundary" };

/**
 * Accessible-reorder keydown resolution: Space/Enter toggles grab, Escape
 * cancels a grab, arrows move the active item (the grabbed one when a grab
 * is active, else the focused one) and report boundaries.
 */
export function listReorderKeyIntent(
  key: string,
  index: number,
  grabbedIndex: number | null,
  itemCount: number,
): ListReorderIntent | null {
  if (key === " " || key === "Enter") {
    return grabbedIndex === index ? { type: "drop" } : { type: "grab" };
  }

  if (key === "Escape" && grabbedIndex !== null) {
    return { type: "cancelGrab" };
  }

  if (key !== "ArrowUp" && key !== "ArrowDown") {
    return null;
  }

  const activeIndex = grabbedIndex ?? index;
  const targetIndex = key === "ArrowUp" ? activeIndex - 1 : activeIndex + 1;

  if (targetIndex < 0 || targetIndex >= itemCount) {
    return { type: "boundary" };
  }

  return { type: "move", from: activeIndex, to: targetIndex };
}

File: packages/contracts/headless/src/edit.rs
//! EditableLabel behavior machine. Mirror of core `edit.ts`;
//! contract: docs/contracts/components/editable-label.md.

fn is_trim_scalar(ch: char) -> bool {
    matches!(
        ch,
        '\u{0009}'..='\u{000D}'
            | '\u{0020}'
            | '\u{0085}'
            | '\u{00A0}'
            | '\u{1680}'
            | '\u{2000}'..='\u{200A}'
            | '\u{2028}'
            | '\u{2029}'
            | '\u{202F}'
            | '\u{205F}'
            | '\u{3000}'
            | '\u{FEFF}'
    )
}

/// Drop the longest prefix and suffix in portable set T. Interior scalars stay.
pub fn trim_editable_label(value: &str) -> String {
    let chars: Vec<char> = value.chars().collect();
    let mut start = 0;
    let mut end = chars.len();
    while start < end && is_trim_scalar(chars[start]) {
        start += 1;
    }
    while end > start && is_trim_scalar(chars[end - 1]) {
        end -= 1;
    }
    chars[start..end].iter().collect()
}

/// Keep at most `max_length` Unicode scalar values.
pub fn clamp_editable_label_draft(draft: &str, max_length: Option<usize>) -> String {
    let Some(limit) = max_length else {
        return draft.to_string();
    };
    draft.chars().take(limit).collect()
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct EditLabelContext {
    pub value: String,
    pub draft: String,
    pub disabled: bool,
    pub max_length: Option<usize>,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum EditLabelState {
    View,
    Editing,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub enum EditLabelEvent {
    StartEdit,
    SetDraft { draft: String },
    Commit,
    CommitBlur,
    Cancel,
    ReplaceValue { value: String },
    SetDisabled { disabled: bool },
    Teardown,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub enum EditLabelEffect {
    EmitEditStart,
    FocusInput,
    EmitCommit {
        value: String,
        previous_value: String,
        restore_focus: bool,
    },
    EmitCancel {
        restore_focus: bool,
    },
}

pub fn edit_label_transition(
    state: EditLabelState,
    context: EditLabelContext,
    event: EditLabelEvent,
) -> (EditLabelState, EditLabelContext, Vec<EditLabelEffect>) {
    match event {
        EditLabelEvent::StartEdit => {
            if state != EditLabelState::View || context.disabled {
                return (state, context, vec![]);
            }
            let next = EditLabelContext {
                draft: context.value.clone(),
                ..context
            };
            (
                EditLabelState::Editing,
                next,
                vec![EditLabelEffect::EmitEditStart, EditLabelEffect::FocusInput],
            )
        }
        EditLabelEvent::SetDraft { draft } => {
            if state != EditLabelState::Editing {
                return (state, context, vec![]);
            }
            let draft = clamp_editable_label_draft(&draft, context.max_length);
            (state, EditLabelContext { draft, ..context }, vec![])
        }
        EditLabelEvent::Commit => commit_from_editing(state, context, true),
        EditLabelEvent::CommitBlur => commit_from_editing(state, context, false),
        EditLabelEvent::Cancel => cancel_from_editing(state, context, true),
        EditLabelEvent::ReplaceValue { value } => {
            if state != EditLabelState::Editing {
                return (
                    state,
                    EditLabelContext {
                        draft: value.clone(),
                        value,
                        ..context
                    },
                    vec![],
                );
            }
            if value == context.value {
                return (state, context, vec![]);
            }
            (
                EditLabelState::View,
                EditLabelContext {
                    draft: value.clone(),
                    value,
                    ..context
                },
                vec![],
            )
        }
        EditLabelEvent::SetDisabled { disabled } => {
            if state == EditLabelState::Editing && disabled {
                let next = EditLabelContext {
                    disabled: true,
                    draft: context.value.clone(),
                    ..context
                };
                return (
                    EditLabelState::View,
                    next,
                    vec![EditLabelEffect::EmitCancel {
                        restore_focus: false,
                    }],
                );
            }
            (
                state,
                EditLabelContext {
                    disabled,
                    ..context
                },
                vec![],
            )
        }
        EditLabelEvent::Teardown => {
            if state != EditLabelState::Editing {
                return (state, context, vec![]);
            }
            let next = EditLabelContext {
                draft: context.value.clone(),
                ..context
            };
            (EditLabelState::View, next, vec![])
        }
    }
}

#[cfg(test)]
mod tests {
    use super::trim_editable_label;

    #[test]
    fn portable_trim_drops_nel_and_bom_that_str_trim_does_not_both_drop() {
        let raw = "\u{0085}Take\u{FEFF}";
        assert_eq!(trim_editable_label(raw), "Take");
        assert_ne!(raw.trim(), "Take");
        assert_eq!(
            trim_editable_label("\u{200B}Keep\u{200B}"),
            "\u{200B}Keep\u{200B}"
        );
    }
}

fn commit_from_editing(
    state: EditLabelState,
    context: EditLabelContext,
    restore_focus: bool,
) -> (EditLabelState, EditLabelContext, Vec<EditLabelEffect>) {
    if state != EditLabelState::Editing {
        return (state, context, vec![]);
    }
    let value = trim_editable_label(&context.draft);
    let previous_value = context.value.clone();
    let next = EditLabelContext {
        draft: context.value.clone(),
        ..context
    };
    (
        EditLabelState::View,
        next,
        vec![EditLabelEffect::EmitCommit {
            value,
            previous_value,
            restore_focus,
        }],
    )
}

fn cancel_from_editing(
    state: EditLabelState,
    context: EditLabelContext,
    restore_focus: bool,
) -> (EditLabelState, EditLabelContext, Vec<EditLabelEffect>) {
    if state != EditLabelState::Editing {
        return (state, context, vec![]);
    }
    let next = EditLabelContext {
        draft: context.value.clone(),
        ..context
    };
    (
        EditLabelState::View,
        next,
        vec![EditLabelEffect::EmitCancel { restore_focus }],
    )
}
